targets/TARGET_Freescale/usb/USBPhy_Kinetis.cpp

```diff
@@ -0,0 +1,638 @@
+    uint16_t  byte_count; // BD[16:32]
+    uint32_t  address;    // Addr
+} BDT;
+
+typedef enum {
+    CTRL_XFER_READY,
+    CTRL_XFER_IN,
+    CTRL_XFER_NONE,
+    CTRL_XFER_OUT
+} ctrl_xfer_t;
+
+// there are:
+//    * 4 bidirectionnal endpt -> 8 physical endpt
+//    * as there are ODD and EVEN buffer -> 8*2 bdt
+MBED_ALIGN(512) BDT bdt[NUMBER_OF_PHYSICAL_ENDPOINTS * 2];  // 512 bytes aligned!
+
+uint8_t * endpoint_buffer[NUMBER_OF_PHYSICAL_ENDPOINTS * 2];
+uint8_t ep0_buffer[2][MAX_PACKET_SIZE_EP0];
+uint8_t ep1_buffer[2][MAX_PACKET_SIZE_EP1];
+uint8_t ep2_buffer[2][MAX_PACKET_SIZE_EP2];
+uint8_t ep3_buffer[2][MAX_PACKET_SIZE_EP3];
+
+static uint8_t set_addr = 0;
+static uint8_t addr = 0;
+static ctrl_xfer_t ctrl_xfer = CTRL_XFER_READY;
+
+static uint32_t Data1  = 0x55555555;
+
+static uint32_t frameNumber() {
+    return((USB0->FRMNUML | (USB0->FRMNUMH << 8)) & 0x07FF);
+}
+
+USBPhy *get_usb_phy()
+{
+    static USBPhyHw usbphy;
+    return &usbphy;
+}
+
+USBPhyHw::USBPhyHw()
+{
+
+}
+
+USBPhyHw::~USBPhyHw()
+{
+}
+
+void USBPhyHw::init(USBPhyEvents *events)
+{
+    this->events = events;
+
+    // Disable IRQ
+    NVIC_DisableIRQ(USB0_IRQn);
+
+#if (defined(FSL_FEATURE_SOC_MPU_COUNT) && (FSL_FEATURE_SOC_MPU_COUNT > 0U))
+    MPU->CESR=0;
+#endif
+
+#if defined(TARGET_KL43Z) || defined(TARGET_K22F) || defined(TARGET_K64F)
+    // enable USBFS clock
+    CLOCK_EnableUsbfs0Clock(kCLOCK_UsbSrcIrc48M, 48000000U);
+#else
+    // choose usb src as PLL
+    SIM->SOPT2 &= ~SIM_SOPT2_PLLFLLSEL_MASK;
+    SIM->SOPT2 |= (SIM_SOPT2_USBSRC_MASK | (1 << SIM_SOPT2_PLLFLLSEL_SHIFT));
+
+    // enable OTG clock
+    SIM->SCGC4 |= SIM_SCGC4_USBOTG_MASK;
+#endif
+
+    // Attach IRQ
+    instance = this;
+
+    // USB Module Configuration
+    // Set BDT Base Register
+    USB0->BDTPAGE1 = (uint8_t)((uint32_t)bdt>>8);
+    USB0->BDTPAGE2 = (uint8_t)((uint32_t)bdt>>16);
+    USB0->BDTPAGE3 = (uint8_t)((uint32_t)bdt>>24);
+
+    // Clear interrupt flag
+    USB0->ISTAT = 0xff;
+
+    // USB Interrupt Enablers
+    USB0->INTEN |= USB_INTEN_TOKDNEEN_MASK |
+                   USB_INTEN_ERROREN_MASK  |
+                   USB_INTEN_USBRSTEN_MASK;
+
+    // Disable weak pull downs
+    USB0->USBCTRL &= ~(USB_USBCTRL_PDE_MASK | USB_USBCTRL_SUSP_MASK);
+
+    USB0->USBTRC0 |= 0x40;
+
+    /* Allocate control endpoint buffers */
+    endpoint_buffer[EP_BDT_IDX(0, TX, ODD)] = ep0_buffer[TX];
+    endpoint_buffer[EP_BDT_IDX(0, RX, ODD)] = ep0_buffer[RX];
+    endpoint_buffer[EP_BDT_IDX(1, TX, ODD)] = ep1_buffer[TX];
+    endpoint_buffer[EP_BDT_IDX(1, RX, ODD)] = ep1_buffer[RX];
+    endpoint_buffer[EP_BDT_IDX(2, TX, ODD)] = ep2_buffer[TX];
+    endpoint_buffer[EP_BDT_IDX(2, RX, ODD)] = ep2_buffer[RX];
+    endpoint_buffer[EP_BDT_IDX(3, TX, ODD)] = ep3_buffer[TX];
+    endpoint_buffer[EP_BDT_IDX(3, RX, ODD)] = ep3_buffer[RX];
+
+    NVIC_SetVector(USB0_IRQn, (uint32_t)&_usbisr);
+    NVIC_EnableIRQ(USB0_IRQn);
+}
+
+void USBPhyHw::deinit()
```

Comment:
shouldn't we add disconnecting code here?
similar to LPC176X implementation

Reply:
Good catch. It has been added.